core/chains/evm/gas/models.go

```diff
@@ -271,12 +271,25 @@ func (e *evmFeeEstimator) GetFee(ctx context.Context, calldata []byte, feeLimit
 		fee.DynamicTipCap = dynamicFee.TipCap
 		return
 	}
-
 	// get legacy fee
 	fee.Legacy, chainSpecificFeeLimit, err = e.EvmEstimator.GetLegacyGas(ctx, calldata, feeLimit, maxFeePrice, opts...)
 	if err != nil {
 		return
 	}
+
+	if e.geCfg.CostMax() != nil {
+		var totalGasCost uint64
+		totalGasCost, err = commonfee.ApplyMultiplier(fee.Legacy.ToInt().Uint64(), e.geCfg.LimitMultiplier())
+		if err != nil {
+			return
+		}
+		maxGasCost := e.geCfg.CostMax().ToInt().Uint64()
+		if totalGasCost > maxGasCost {
+			err = fmt.Errorf("estimated gas cost %d exceeds maximum allowed gas cost %d", totalGasCost, maxGasCost)
```

Comment:
In my opinion, we shouldn't return error. Instead, we should return the fee that keeps the gas cost equal to the max cost.

So newfee = maxGasCost / e.geCfg.LimitMultiplier()

And log a warning that we reached max gas cost for this tx, thus capping fees to this max value.
And return this newfee.

Reply:
got it. PTAL